connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/ReflectionScanner.java

```diff
@@ -77,53 +79,59 @@ private static <T> String versionFor(Class<? extends T> pluginKlass) throws Refl
-        );
+        Set<URL> urls = new HashSet<>();
+        Collections.addAll(urls, source.urls());
+        ClassGraph classGraphBuilder = new ClassGraph()
+                .addClassLoader(source.loader())
+                .filterClasspathElementsByURL(urls::contains)
+                .enableExternalClasses()
+                .enableAllInfo();
+        try (ScanResult classGraph = classGraphBuilder.scan()) {
+            return new PluginScanResult(
+                  getPluginDesc(classGraph, PluginType.SINK, source),
+                  getPluginDesc(classGraph, PluginType.SOURCE, source),
+                  getPluginDesc(classGraph, PluginType.CONVERTER, source),
+                  getPluginDesc(classGraph, PluginType.HEADER_CONVERTER, source),
+                  getTransformationPluginDesc(source, classGraph),
+                  getPredicatePluginDesc(source, classGraph),
+                  getServiceLoaderPluginDesc(PluginType.CONFIGPROVIDER, source),
+                  getServiceLoaderPluginDesc(PluginType.REST_EXTENSION, source),
+                  getServiceLoaderPluginDesc(PluginType.CONNECTOR_CLIENT_CONFIG_OVERRIDE_POLICY, source)
+          );
+        }
     }
 
     @SuppressWarnings({"unchecked"})
-    private SortedSet<PluginDesc<Predicate<?>>> getPredicatePluginDesc(PluginSource source, Reflections reflections) {
-        return (SortedSet<PluginDesc<Predicate<?>>>) (SortedSet<?>) getPluginDesc(reflections, PluginType.PREDICATE, source);
+    private SortedSet<PluginDesc<Predicate<?>>> getPredicatePluginDesc(PluginSource source, ScanResult classGraph) {
+        return (SortedSet<PluginDesc<Predicate<?>>>) (SortedSet<?>) getPluginDesc(classGraph, PluginType.PREDICATE, source);
     }
 
     @SuppressWarnings({"unchecked"})
-    private SortedSet<PluginDesc<Transformation<?>>> getTransformationPluginDesc(PluginSource source, Reflections reflections) {
-        return (SortedSet<PluginDesc<Transformation<?>>>) (SortedSet<?>) getPluginDesc(reflections, PluginType.TRANSFORMATION, source);
+    private SortedSet<PluginDesc<Transformation<?>>> getTransformationPluginDesc(PluginSource source, ScanResult classGraph) {
+        return (SortedSet<PluginDesc<Transformation<?>>>) (SortedSet<?>) getPluginDesc(classGraph, PluginType.TRANSFORMATION, source);
     }
 
     @SuppressWarnings({"unchecked"})
     private <T> SortedSet<PluginDesc<T>> getPluginDesc(
-            Reflections reflections,
+            ScanResult classGraph,
             PluginType type,
             PluginSource source
     ) {
-        Set<Class<? extends T>> plugins;
+        ClassInfoList plugins;
+        Class<T> kclass = (Class<T>) type.superClass();
         try {
-            plugins = reflections.getSubTypesOf((Class<T>) type.superClass());
-        } catch (ReflectionsException e) {
+            plugins = classGraph.getSubclasses(kclass.getName());
+            if (plugins.isEmpty()) {
```

Comment:
nit: You could check kclass.isInterface() first rather than calling the wrong method first.

Reply:
Done.